src/FreeAge/common/type_info.cpp

```diff
@@ -0,0 +1,621 @@
+
+#include "FreeAge/common/free_age.hpp"
+#include "FreeAge/common/unit_types.hpp"
+#include "FreeAge/common/building_types.hpp"
+#include "FreeAge/common/player.hpp"
+#include "FreeAge/common/damage.hpp"
+#include "FreeAge/common/resources.hpp"
+#include "FreeAge/common/object_types.hpp"
+
+// Naming conversions:
+// *Stats: collection of values
+// *Info: the base *Stats, a list of modications on it and the resulting *Stats
+
+// Player has a CivilizationInfo, a vector<UnitTypeInfo>, a vector<BuildingTypeInfo> and a vector<TechnologyInfo>
+// The game data provide the base UnitTypeStats, BuildingTypeStats, TechnologyStats and CivilizationStats
+
+// Note: The *Stats object do NOT contain information about the dependencies and building's interfaces (unit production,
+//       technology research). The consept of the *Stats and *Info is to have the needed information by the game logic
+//       in a ready to use *Stats struct, but also be able to provide to the gui information about the how the values
+//       from the staring base *Stats have been calculated (by keep track of all the modications).
+//       Dependencies and building's interfaces are accessed by different parts of the game logic and also there is no
+//       need to know anything about the previous state.
+
+// General structure
+//
+// Player
+//  \ CivilizationInfo
+//  \ vector<UnitTypeInfo>
+//  \ vector<BuildingTypeInfo>
+//  \ vector<TechnologyInfo>
+//
+// XxInfo
+// \ XxStats baseStats                        (provided by the game data)
+// \ std::vector<Modification> modifications  (starts empty)
+// \ XxStats stats                            (updates every time the modifications change)
+//
+// XxStats
+// \ raw values
+//
+// Modification
+// \ Type        (what to modify)   (TODO: rename to Target?)
+// \ Operation   (how to modify it)
+// \ int value                (with two ints everything is possible)
+// \ int extra
+//
+
+struct UnitTypeInfo {
+
+  UnitTypeInfo(UnitType type, UnitTypeStats baseStats)
+      : type(type),
+        baseStats(baseStats) {
+    stats = baseStats;
+  }
+
+  UnitType type;
+  UnitTypeStats baseStats;   // immutable
+  UnitTypeStats stats;       // mutable
+
+  std::vector<Modification> modifications;
+
+  void AddModification(Modification mod) {
+
+    bool samePriorityAsLast = modifications.back().priority = mod.priority;
+    if (samePriorityAsLast) {
+      // this opt is not so important, could be removed
+      modifications.push_back(mod);
+      mod.ApplyToUnit(&stats, &baseStats);
+
+    } else {
+      modifications.push_back(mod); // TODO: instert at the correct position based on the priority
+
+      // re-eval from the start
+      stats = baseStats;
+      for (auto i : modifications) {
+        i.ApplyToUnit(&stats, &baseStats);
+      }
+    }
+  }
+
+  void MakeCurrentStastTheBaseStats() { // for converting (in order to hide the info about the opponent upgrades)
```

Comment:
I'm not sure either, but could not inferred 100%, you could see that it has more armor but not from which tech it got.

Reply:
Is it important to hide which technology a change came from? It seems to me like there are probably only very few examples of where this is even ambiguous (unless I am overlooking something ... but I am a mediocre player only, so that might be the case ;) ).

It seems to me that the best way for the server to tell the clients about technologies (minimizing the required bandwidth) could be to simply pass on which ones have been researched, as soon as a unit/building is visible to which the effect applies (the clients could then apply the effects on their side based on their knowledge about the technologies). This would however make that information available about the opponents as well (even if the standard client would not display it, it would be easy to modify the client to do that).

Reply:
I was under the wrong impression that you could not see the upgrade values of enemy units (I think because of Dota 2???).

So now its much more simpler. Yes, I agree passing around the techs instead of the stats is the way to go.

Reply:
> Not sure whether it is actually hidden or just frozen

Now I understand...